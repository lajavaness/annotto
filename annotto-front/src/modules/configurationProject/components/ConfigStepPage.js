import { Form, Input, Select } from 'antd'
import { debounce, isEmpty } from 'lodash'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import React, { useCallback, useEffect, useMemo } from 'react'
import moment from 'moment'

import { fetchClients, fetchUsers } from 'modules/root/actions/rootActions'
import { updateConfigProject } from 'modules/configurationProject/actions/configurationProjectActions'

import {
	selectConfigProject,
	selectInitialConfigDefaultTag,
} from 'modules/configurationProject/selectors/configurationProjectSelectors'
import { selectProjectClients, selectUsers } from 'modules/root/selectors/rootSelectors'

import {
	ADMINS,
	CLIENT,
	DATASCIENTISTS,
	DEADLINE,
	DESCRIPTION,
	NAME,
	PROJECT,
	TAGS,
	TYPE,
	USERS,
} from 'shared/enums/configurationsProjectPropertiesTypes'
import projectType from 'shared/enums/projectType'
import tagTypes from 'shared/enums/tagTypes'

import * as Styled from 'modules/configurationProject/components/__styles__/ConfigStepPage.styles'

const ConfigStepPage = () => {
	const projectConfig = useSelector(selectConfigProject)
	const users = useSelector(selectUsers)
	const client = useSelector(selectProjectClients)
	const defaultInitialTags = useSelector(selectInitialConfigDefaultTag)

	const dispatch = useDispatch()

	const { t } = useTranslation('configurationProject')

	const [form] = Form.useForm()

	const projectConfigWithDefaultTags = useMemo(
		() => ({
			...projectConfig,
			defaultTags: tagTypes.map((tag) => t(`configurationProject:config.${tag}`)),
		}),
		[projectConfig, t]
	)

	const tagOptions = useMemo(
		() => [
			...(!isEmpty(defaultInitialTags)
				? defaultInitialTags
						.filter((value) => !tagTypes.some((tag) => value !== tag))
						.map((value) => ({ value, label: value }))
				: []),
			...tagTypes.map((value) => ({
				value: t(`configurationProject:config.${value}`),
				label: t(`configurationProject:config.${value}`),
			})),
		],
		[defaultInitialTags, t]
	)

	const projectOptions = projectType.map((value) => ({ value, label: value }))
	const userOptions = users?.map(({ email }) => ({ value: email, label: email }))
	const clientOptions = client?.data?.map(({ name }) => ({ value: name, label: name }))
	const today = useMemo(() => moment(), [])

	const disabledDate = useCallback((current) => current.startOf('day').diff(today.startOf('day'), 'days') < 0, [today])

	useEffect(() => {
		if (projectConfig) {
			form.setFieldsValue(projectConfig)
		}
	}, [form, projectConfig])

	const _onUserSearch = useCallback(
		debounce((value) => dispatch(fetchUsers(value)), 500),
		[dispatch]
	)

	const _onClientSearch = useCallback(
		debounce((value) => dispatch(fetchClients(value)), 500),
		[dispatch]
	)

	const _onClientFocus = useCallback(() => dispatch(fetchClients()), [dispatch])

	const _onValuesChange = useCallback(
		debounce((_, values) => dispatch(updateConfigProject(PROJECT, values)), 500),
		[dispatch]
	)

	return (
		<Styled.Root>
			<Styled.Form
				layout="vertical"
				initialValues={projectConfigWithDefaultTags}
				form={form}
				onValuesChange={_onValuesChange}
			>
				<Styled.ItemsContainer>
					<Styled.Title>{t('configurationProject:config.detail')}</Styled.Title>
					<Form.Item required label={t('configurationProject:config.client')} name={CLIENT} onFocus={_onClientFocus}>
						<Styled.AutoComplete onSearch={_onClientSearch} options={clientOptions} />
					</Form.Item>
					<Form.Item required label={t('configurationProject:config.type')} name={TYPE}>
						<Styled.Select options={projectOptions} />
					</Form.Item>
					<Form.Item required label={t('configurationProject:config.name')} name={NAME}>
						<Styled.Input />
					</Form.Item>
					<Form.Item required label={t('configurationProject:config.deadline')} name={DEADLINE}>
						<Styled.DatePicker disabledDate={disabledDate} format={'L'} />
					</Form.Item>
					<Form.Item required label={t('configurationProject:config.description')} name={DESCRIPTION}>
						<Input.TextArea autoSize={{ minRows: 4 }} />
					</Form.Item>
					<Form.Item label={t('configurationProject:config.tags')} name={TAGS}>
						<Select mode="tags" options={tagOptions} />
					</Form.Item>
					<Styled.Title>{t('configurationProject:config.permissions')}</Styled.Title>
					<Form.Item label={t('configurationProject:config.admins')} name={ADMINS}>
						<Select mode="multiple" onSearch={_onUserSearch} options={userOptions} />
					</Form.Item>
					<Form.Item label={t('configurationProject:config.dataScientists')} name={DATASCIENTISTS}>
						<Select mode="multiple" onSearch={_onUserSearch} options={userOptions} />
					</Form.Item>
					<Form.Item label={t('configurationProject:config.users')} name={USERS}>
						<Select mode="multiple" onSearch={_onUserSearch} options={userOptions} />
					</Form.Item>
				</Styled.ItemsContainer>
			</Styled.Form>
		</Styled.Root>
	)
}

export default ConfigStepPage
