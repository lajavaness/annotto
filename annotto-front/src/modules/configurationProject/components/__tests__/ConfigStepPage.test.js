import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'
import { createStore } from 'redux-dynamic-modules'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import ConfigStepPage from 'modules/configurationProject/components/ConfigStepPage'

const initialState = {}
const getStore = (state) => createStore({ initialState: state })
const getInstance = (state = initialState) => (
	<Provider store={getStore(state)}>
		<ThemeProvider theme={theme}>
			<ConfigStepPage />
		</ThemeProvider>
	</Provider>
)

describe('ConfigStepPage', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
