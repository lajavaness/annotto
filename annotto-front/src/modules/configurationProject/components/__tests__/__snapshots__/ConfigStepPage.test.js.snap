// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigStepPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsBqHK ljdoNJ"
  >
    <form
      class="ant-form ant-form-vertical sc-kmATbt jcJPbe"
    >
      <div
        class="sc-cvJIAo iEPlST"
      >
        <div
          class="sc-aKXVD hQbhhA"
        >
          configurationProject:config.detail
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="client"
              title="configurationProject:config.client"
            >
              configurationProject:config.client
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-auto-complete sc-fybuyZ iHdAvI ant-select-single ant-select-show-search"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="client_list_0"
                        aria-autocomplete="list"
                        aria-controls="client_list"
                        aria-haspopup="listbox"
                        aria-owns="client_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="client"
                        role="combobox"
                        type="search"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-placeholder"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="type"
              title="configurationProject:config.type"
            >
              configurationProject:config.type
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select sc-gIRjGK fZPFMV ant-select-single ant-select-show-arrow"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="type_list_0"
                        aria-autocomplete="list"
                        aria-controls="type_list"
                        aria-haspopup="listbox"
                        aria-owns="type_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="type"
                        readonly=""
                        role="combobox"
                        style="opacity: 0;"
                        type="search"
                        unselectable="on"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-placeholder"
                    />
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="name"
              title="configurationProject:config.name"
            >
              configurationProject:config.name
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input sc-dkmKIT ifXgxT"
                  id="name"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="deadline"
              title="configurationProject:config.deadline"
            >
              configurationProject:config.deadline
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-picker sc-iWFTwQ eBGXRn"
                >
                  <div
                    class="ant-picker-input"
                  >
                    <input
                      autocomplete="off"
                      id="deadline"
                      placeholder="Select date"
                      readonly=""
                      size="12"
                      title=""
                      value=""
                    />
                    <span
                      class="ant-picker-suffix"
                    >
                      <span
                        aria-label="calendar"
                        class="anticon anticon-calendar"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="calendar"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="description"
              title="configurationProject:config.description"
            >
              configurationProject:config.description
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <textarea
                  class="ant-input"
                  id="description"
                  style="height: 0px; min-height: -16px; max-height: 9007199254740991px; overflow-y: hidden; resize: none; overflow-x: hidden;"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="defaultTags"
              title="configurationProject:config.tags"
            >
              configurationProject:config.tags
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-multiple ant-select-show-search"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <div
                      class="ant-select-selection-overflow"
                    >
                      <div
                        class="ant-select-selection-overflow-item"
                        style="opacity: 1;"
                      >
                        <span
                          class="ant-select-selection-item"
                        >
                          <span
                            class="ant-select-selection-item-content"
                          >
                            configurationProject:config.delete
                          </span>
                          <span
                            aria-hidden="true"
                            class="ant-select-selection-item-remove"
                            style="user-select: none;"
                            unselectable="on"
                          >
                            <span
                              aria-label="close"
                              class="anticon anticon-close"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="close"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                      <div
                        class="ant-select-selection-overflow-item"
                        style="opacity: 1;"
                      >
                        <span
                          class="ant-select-selection-item"
                        >
                          <span
                            class="ant-select-selection-item-content"
                          >
                            configurationProject:config.check
                          </span>
                          <span
                            aria-hidden="true"
                            class="ant-select-selection-item-remove"
                            style="user-select: none;"
                            unselectable="on"
                          >
                            <span
                              aria-label="close"
                              class="anticon anticon-close"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="close"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                      <div
                        class="ant-select-selection-overflow-item"
                        style="opacity: 1;"
                      >
                        <span
                          class="ant-select-selection-item"
                        >
                          <span
                            class="ant-select-selection-item-content"
                          >
                            configurationProject:config.discuss
                          </span>
                          <span
                            aria-hidden="true"
                            class="ant-select-selection-item-remove"
                            style="user-select: none;"
                            unselectable="on"
                          >
                            <span
                              aria-label="close"
                              class="anticon anticon-close"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="close"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                      <div
                        class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                        style="opacity: 1;"
                      >
                        <div
                          class="ant-select-selection-search"
                          style="width: 0px;"
                        >
                          <input
                            aria-activedescendant="defaultTags_list_0"
                            aria-autocomplete="list"
                            aria-controls="defaultTags_list"
                            aria-haspopup="listbox"
                            aria-owns="defaultTags_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="defaultTags"
                            role="combobox"
                            type="search"
                            value=""
                          />
                          <span
                            aria-hidden="true"
                            class="ant-select-selection-search-mirror"
                          >
                             
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-aKXVD hQbhhA"
        >
          configurationProject:config.permissions
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="admins"
              title="configurationProject:config.admins"
            >
              configurationProject:config.admins
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-multiple ant-select-show-search"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <div
                      class="ant-select-selection-overflow"
                    >
                      <div
                        class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                        style="opacity: 1;"
                      >
                        <div
                          class="ant-select-selection-search"
                          style="width: 0px;"
                        >
                          <input
                            aria-activedescendant="admins_list_0"
                            aria-autocomplete="list"
                            aria-controls="admins_list"
                            aria-haspopup="listbox"
                            aria-owns="admins_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="admins"
                            readonly=""
                            role="combobox"
                            style="opacity: 0;"
                            type="search"
                            unselectable="on"
                            value=""
                          />
                          <span
                            aria-hidden="true"
                            class="ant-select-selection-search-mirror"
                          >
                             
                          </span>
                        </div>
                      </div>
                    </div>
                    <span
                      class="ant-select-selection-placeholder"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="dataScientists"
              title="configurationProject:config.dataScientists"
            >
              configurationProject:config.dataScientists
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-multiple ant-select-show-search"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <div
                      class="ant-select-selection-overflow"
                    >
                      <div
                        class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                        style="opacity: 1;"
                      >
                        <div
                          class="ant-select-selection-search"
                          style="width: 0px;"
                        >
                          <input
                            aria-activedescendant="dataScientists_list_0"
                            aria-autocomplete="list"
                            aria-controls="dataScientists_list"
                            aria-haspopup="listbox"
                            aria-owns="dataScientists_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="dataScientists"
                            readonly=""
                            role="combobox"
                            style="opacity: 0;"
                            type="search"
                            unselectable="on"
                            value=""
                          />
                          <span
                            aria-hidden="true"
                            class="ant-select-selection-search-mirror"
                          >
                             
                          </span>
                        </div>
                      </div>
                    </div>
                    <span
                      class="ant-select-selection-placeholder"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="users"
              title="configurationProject:config.users"
            >
              configurationProject:config.users
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-multiple ant-select-show-search"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <div
                      class="ant-select-selection-overflow"
                    >
                      <div
                        class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                        style="opacity: 1;"
                      >
                        <div
                          class="ant-select-selection-search"
                          style="width: 0px;"
                        >
                          <input
                            aria-activedescendant="users_list_0"
                            aria-autocomplete="list"
                            aria-controls="users_list"
                            aria-haspopup="listbox"
                            aria-owns="users_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="users"
                            readonly=""
                            role="combobox"
                            style="opacity: 0;"
                            type="search"
                            unselectable="on"
                            value=""
                          />
                          <span
                            aria-hidden="true"
                            class="ant-select-selection-search-mirror"
                          >
                             
                          </span>
                        </div>
                      </div>
                    </div>
                    <span
                      class="ant-select-selection-placeholder"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
